src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+          break;
+        }
+
+        case kind::BITVECTOR_UDIV_TOTAL:
+        case kind::BITVECTOR_OR:
+        case kind::BITVECTOR_NOR:
+        case kind::BITVECTOR_XOR:
+        case kind::BITVECTOR_XNOR:
+        case kind::BITVECTOR_AND:
+        case kind::BITVECTOR_NAND:
+        case kind::BITVECTOR_SHL:
+        case kind::BITVECTOR_LSHR:
+        case kind::BITVECTOR_ASHR:
+        {
+          unsigned wmax = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] > wmax) wmax = visited[n[i]];
+          visited[n] = wmax;
+          break;
+        }
+
+        case kind::BITVECTOR_UREM_TOTAL:
+        {
+          visited[n] = visited[n[1]];
+          break;
+        }
+
+        case kind::BITVECTOR_PLUS:
+        case kind::BITVECTOR_SUB:
+        {
+          unsigned w = 0, wmax = visited[n[0]];
+          for (unsigned i = 1, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] > wmax) wmax = visited[n[i]];
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            if (visited[n[i]] >= wmax) w += 1;
+          w += wmax - 1;
+          if (w > utils::getSize(n)) return 0; /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_NOT:
+        case kind::BITVECTOR_NEG:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        default:
+        {
+          visited[n] = utils::getSize(n);
+        }
+      }
+    }
+  }
+  return visited[expr];
+}
+
+BVGaussElim::Result BVGaussElim::gaussElim(Integer prime,
+                                           const vector<Integer> &rhs,
+                                           const vector<vector<Integer>> &lhs,
+                                           vector<Integer> &resrhs,
+                                           vector<vector<Integer>> &reslhs)
+
+{
+  Assert(prime > 0);
+  Assert(lhs.size());
+  Assert(lhs.size() == rhs.size());
+  Assert(lhs.size() <= lhs[0].size());
+
+  /* special case: zero ring */
+  if (prime == 1)
+  {
+    resrhs = vector<Integer>(rhs.size(), Integer(0));
+    reslhs = vector<vector<Integer>>(
+        lhs.size(), vector<Integer>(lhs[0].size(), Integer(0)));
+    return BVGaussElim::Result::UNIQUE;
+  }
+
+  resrhs = vector<Integer>(rhs);
+  reslhs = vector<vector<Integer>>(lhs);
+
+  size_t nrows = lhs.size();
+  size_t ncols = lhs[0].size();
+
+#ifdef CVC4_ASSERTIONS
+  for (size_t i = 1; i < nrows; ++i) Assert(lhs[i].size() == ncols);
+#endif
+  /* (1) if element in pivot column is non-zero and != 1, divide row elements
+   *     by element in pivot column modulo prime, i.e., multiply row with
+   *     multiplicative inverse of element in pivot column modulo prime
+   *
+   * (2) subtract pivot row from all rows below pivot row
+   *
+   * (3) subtract (multiple of) current row from all rows above s.t. all
+   *     elements in current pivot column of above rows equal to one
```

Comment:
"of above rows equal to one" does not quite parse

Reply:
It seems like this comment hasn't changed :)

Reply:
It did, see 
https://github.com/CVC4/CVC4/commit/fc8ecfe96e9c662e1ed473e8c6b9a72cc5e8bf42#diff-b68f14173b13a57846b56cca910f52e0L203.

Reply:
Hmm, still difficult to parse IMHO. "all elements in current pivot column of above rows equal to one" -> "all elements in pivot column above current row become equal to one"

Reply:
That's better, yes.